File: solutions/typescript/030-sudoku-solver.ts
/**
 * Problem 30: Sudoku Solver (LeetCode 37)
 * Difficulty: Hard
 * Language: TypeScript
 */
function solveSudoku(board: string[][]): void {
  const digits = '123456789'.split('');

  const isValid = (board: string[][], r: number, c: number, d: string): boolean =>
    board[r].every(v => v !== d) &&
    board.every(row => row[c] !== d) &&
    board.slice(Math.floor(r/3)*3, Math.floor(r/3)*3+3)
      .every(row => row.slice(Math.floor(c/3)*3, Math.floor(c/3)*3+3)
        .every(v => v !== d));

  const solve = (board: string[][]): string[][] | null => {
    for (let i = 0; i < 9; i++) {
      for (let j = 0; j < 9; j++) {
        if (board[i][j] === '.') {
          for (const d of digits) {
            if (isValid(board, i, j, d)) {
              const newBoard = board.map(r => [...r]);
              newBoard[i][j] = d;
              const result = solve(newBoard);
              if (result) return result;
            }
          }
          return null;
        }
      }
    }
    return board;
  };

  const result = solve(board);
  if (result) result.forEach((row, i) => row.forEach((v, j) => board[i][j] = v));
}

File: solutions/rust/030-sudoku-solver.rs
/**
 * Problem 30: Sudoku Solver (LeetCode 37)
 * Difficulty: Hard
 * Language: Rust
 */
pub fn solve_sudoku(board: &mut Vec<Vec<char>>) {
    fn is_valid(board: &Vec<Vec<char>>, r: usize, c: usize, d: char) -> bool {
        let br = (r / 3) * 3;
        let bc = (c / 3) * 3;
        (0..9).all(|i| board[r][i] != d) &&
        (0..9).all(|i| board[i][c] != d) &&
        (0..3).all(|i| (0..3).all(|j| board[br+i][bc+j] != d))
    }

    fn solve(board: &mut Vec<Vec<char>>) -> bool {
        for i in 0..9 {
            for j in 0..9 {
                if board[i][j] == '.' {
                    for d in '1'..='9' {
                        if is_valid(board, i, j, d) {
                            board[i][j] = d;
                            if solve(board) { return true; }
                            board[i][j] = '.';
                        }
                    }
                    return false;
                }
            }
        }
        true
    }
    solve(board);
}
